File: golang/sliding_window/max_number_of_vowels_in_substring.go
package main

func isVowel(c byte) bool {
	vowels := "aeiouAEIOU"

	for i := 0; i < len(vowels); i++ {
		if c == vowels[i] {
			return true
		}
	}

	return false
}

func maxVowels(s string, k int) int {
	maxVowelCount := 0
	currentVowelCount := 0

	for i := 0; i < k; i++ {
		if isVowel(s[i]) {
			currentVowelCount++
		}
	}

	maxVowelCount = currentVowelCount

	for i := k; i < len(s); i++ {
		if isVowel(s[i]) {
			currentVowelCount++
		}

		if isVowel(s[i-k]) {
			currentVowelCount--
		}

		if currentVowelCount > maxVowelCount {
			maxVowelCount = currentVowelCount
		}
	}

	return maxVowelCount
}

File: java/src/main/java/com/sanedge/interview/sliding_window/MaxNumberOfVowelsInSubString.java
package com.sanedge.interview.sliding_window;

public class MaxNumberOfVowelsInSubString {
  private boolean isVowel(char c) {
    String vowels = "aeiouAEIOU";
    return vowels.indexOf(c) != -1;
  }

  public int maxVowels(String s, int k) {
    int maxVowelCount = 0;
    int currentVowelCount = 0;

    for (int i = 0; i < k; i++) {
      if (isVowel(s.charAt(i))) {
        currentVowelCount++;
      }
    }

    for (int i = k; i < s.length(); i++) {
      if (isVowel(s.charAt(i))) {
        currentVowelCount++;
      }

      if (isVowel(s.charAt(i - k))) {
        currentVowelCount--;
      }

      if (currentVowelCount > maxVowelCount) {
        maxVowelCount = currentVowelCount;
      }
    }

    return maxVowelCount;
  }

}
